contributors/design-proposals/instrumentation/external-metrics-api.md

```diff
@@ -0,0 +1,87 @@
+# **External Metrics API**
+
+# Overview
+
+[HPA v2 API extension proposal](https://github.com/kubernetes/community/blob/hpa_external/contributors/design-proposals/autoscaling/hpa-external-metrics.md) introduces new External metric type for autoscaling based on metrics coming from outside of Kubernetes cluster. This document proposes a new External Metrics API that will be used by HPA controller to get those metrics.
+
+This API performs a similar role to and is based on existing [Custom Metrics API](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md). Unless explicitly specified otherwise all sections related to semantics, implementation and design decisions in [Custom Metrics API design](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md) apply to External Metrics API as well. It is generally expected that a Custom Metrics Adapter will provide both Custom Metrics API and External Metrics API, however, this is not a requirement and both APIs can be implemented and used separately.
+
+
+# API
+
+The API will consist of a single path:
+
+
+```
+/apis/external.metrics.k8s.io/v1beta1/namespaces/<namespace_name>/<metric_name>?labelSelector=<selector>
+```
+
+Similar to endpoints in Custom Metrics API it would only support GET requests.
+
+The query would return the `ExternalMetricValueList` type described below:
+
+```go
+// a list of values for a given metric for some set labels
+type ExternalMetricValueList struct {
+       metav1.TypeMeta `json:",inline"`
+       metav1.ListMeta `json:"metadata,omitempty"`
+
+       // value of the metric matching a given set of labels
+       Items []ExternalMetricValue `json:"items"`
+}
+
+// a metric value for external metric
+type ExternalMetricValue struct {
+    metav1.TypeMeta`json:",inline"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // label set identifying the value within metric
+    MetricLabels map[string]string `json:"metricLabels"`
+
+    // indicates the time at which the metrics were produced
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the window ([Timestamp-Window, Timestamp]) from
+    // which these metrics were calculated, when returning rate
+    // metrics calculated from cumulative metrics (or zero for
+    // non-calculated instantaneous metrics).
+    WindowSeconds *int64 `json:"window,omitempty"`
+
+    // the value of the metric
+    Value resource.Quantity
+}
+```
+
+# Semantics
+
+## Namespaces
+
+Kubernetes namespaces don't have a natural 1-1 mapping to metrics coming from outside of Kubernetes. It is up to adapter implementing the API to decide which metric is available in which namespace. In particular a single metric may be available through many different namespaces.
```

Comment:
I'm not sure I understand why we want this in the first place. One of the problems with the custom metrics API is as you stated that the metric may be completely decoupled from Kubernetes, so why introduce Kubernetes paradigms anyways? Seems like we should just leave this out and leave it up to the user to add any select by labels as they wish.

Reply:
The main point is to allow writing RBAC rules to allow or forbid access to a given metric in a given namespace. I think this is good enough reason to add it even if I expect some (most?) adapters may choose to just ignore this parameter. Of course adapters are free to use it for some custom access control if required.

Reply:
that makes sense :+1: